### ISSUE TITLE
Near-identical weights when performing subsequent min-semivar optimisations

### ISSUE THREAD

**Issue by tuffah12** [robertmartin8/PyPortfolioOpt#337]
Hi Robert (and all others of course too! :) ),

First of all, thanks for your amazing work. It's amazing that you took the effort to develop a package that is of great use to others. 

I'm checking how the length of an estimation window (training set) would affect portfolio allocations using different allocation methods and I'm using PyPortfolioOpt for the min-semivariance allocation.

The issue also arises in this code snippet:

``` python
windows = [t1, t2, t3]
for window in windows:
    sample = prices[window:]
    exp = expected_returns.mean_historical_return(sample)
    s = EfficientSemivariance(exp, sample.pct_change(), benchmark=0)
    s.min_semivariance()
    used_weights = [(ticker, weight) for ticker, weight in zip(s.tickers, s.weights) if weight > 0.001]
    print(used_weights)
    used_tickers = [ticker for ticker, weight in used_weights]
    print(risk_models.semicovariance(sample[used_tickers], benchmark=0).head())
```

`prices` is a pd.DataFrame with adjusted closing prices of most S&P500 firms from jan 2006 - jan 2021. t1= 2006, t2 = 2011 and t3  = 2014 (all dates are january first). 

The last four lines are just to be sure that the semicovariance matrices are not identical (which they aren't).
The first time I ran that loop, I got _slightly_ different weights and they converged to be identical after three or so loops.
I tried putting `del s` in the end of the loop but that wouldn't make a difference.

Stay awesome!

Mark.

edit: my data can be found [here](https://gist.github.com/tuffah12/32424e3320e67177f6196d1e3e8bc798)! :)

**Comment by phschiele** [comment]
Hi @tuffah12, 
I have replicated your code and the issue is that the price dataframe contains NaN values.
You should see the warning `UserWarning: Removing NaNs from returns` while running your code. Essentially, all rows are thrown out which do contain any NaN value. 
The first row where this is true is "2020-03-19" -> No matter the start date, all optimizations are run on the same inputs.
Hope this helps :)
_________
P.S.: when posting multi-line code snippets, it's better to use triple backticks, as one can otherwise not see the line breaks. You can even add 
'''python
your code
'''
to make it look even nicer (of course with backticks).
E.g. something like this:
```python
windows = [t1,t2,t3]
for window in windows:
    sample = prices[window:]
    exp = expected_returns.mean_historical_return(sample)
    s = EfficientSemivariance(exp, sample.pct_change(), benchmark=0)
    s.min_semivariance()
    used_weights = [(ticker, weight) for ticker, weight in zip(s.tickers, s.weights) if weight > 0.001]
    print(used_weights)
    used_tickers = [ticker for ticker, weight in used_weights]
    print(risk_models.semicovariance(sample[used_tickers], benchmark=0).head())
```

**Comment by tuffah12** [comment]
Thanks for the quick response @phschiele. It was my first post so thanks for letting me know!

I ignored that warning as I assumed that for min-semivariance optimisation the only variable that matters would be the semicovariance matrix, isn't it? Anyways, I've adjusted my sampling method and works like a charm now, so thanks a lot! I do notice it takes a lot longer now as it is taking a whole longer sample to work with.
